### ISSUE TITLE
CodeAction previews don't work on "Source" actions

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode#210171]
CodeActions on the lightbulb menu can be previewed with Ctrl+Enter:

![Screenshot 2024-04-11 154344](https://github.com/microsoft/vscode/assets/1078012/ea043fd3-6bb1-4b0c-9c5b-998125183815)

However this isn't the case for "Source" CodeActions:

![image](https://github.com/microsoft/vscode/assets/1078012/ee870a8c-a3e9-40c5-b9b0-f1bc176e9266)

**Comment by DanTup** [comment]
Reading https://github.com/microsoft/vscode/pull/192238, I wonder if the difference here is actually Command vs inline edits?

If so, should extensions be computing all of their edits in `resolve` instead? It seems like this could cause lots of additional computation (because of https://github.com/microsoft/vscode/issues/201871)

**Comment by mjbvz** [comment]
@justschen I updated JS/TS'so organize imports to use `resolveCodeAction` but still don't see this working

**Comment by justschen** [comment]
@mjbvz hmmm i'll do some investigation - i might have a condition or have hooked it up to only quick fix items

**Comment by justschen** [comment]
@DanTup so this is actually related to an issue I'm working on ATM - it's not that preview doesn't work, the label just won't show that you can preview.

if you try doing `shift + enter` while selecting that source action, you will still be able to see the previews come thru

**Comment by DanTup** [comment]
@justschen I tried doing Shift+Enter but it didn't do anything (using TypeScript):

https://github.com/microsoft/vscode/assets/1078012/5d04e481-cc60-4a70-b401-c20ec4c05065


> to avoid the multiple calls to resolve, we could statically determine wheterh code actions would have available previews or not (based on the type of code action)

Can you elaborate on this? What do you mean by "type of action"? Without resolving, how could you know whether there will be edits?

**Comment by justschen** [comment]
@DanTup sorry, it's `CmdCtrl + enter` not `Shift + Enter`. In the cases where there are edits, it should perform fine atm.

> Can you elaborate on this? What do you mean by "type of action"? Without resolving, how could you know whether there will be edits?

Some code actions (Refactor Rewrites, Refactor Inline, Refactor Extract, Moves...) are guaranteed to have edits. For a naive fix, we specify that these code actions (and maybe source kinds) are previewable. 

Still looking into fixing the PR where we resolved early -> determed if we have edits -> came back with the correct label.

**Comment by DanTup** [comment]
>  sorry, it's `CmdCtrl + enter` not Shift + Enter.
> In the cases where there are edits, it should perform fine atm.

Ah, right - so it doesn't show a preview right now for TypeScript - but I presume that's because it's using a command and not inline edits? Is there any plan to be able to support previews for things that use commands? (Or, can the rules around `resolve` be made clearer so it's clear a server should prefer computing things like Organize Imports in resolve and not use commands?).

> Some code actions (Refactor Rewrites, Refactor Inline, Refactor Extract, Moves...) are guaranteed to have edits. For a naive fix, we specify that these code actions (and maybe source kinds) are previewable.

Assuming "guaranteed to have edits" means  inline (and not just use commands), I think this assumption is flawed. Here's the "Extract Method" CodeAction for Dart:

```
{
	"command": {
		"arguments": [
			"EXTRACT_METHOD",
			"C:\\Dev\\Test Projects\\dart_sample\\bin\\main.dart",
			1,
			31,
			23,
			null
		],
		"command": "refactor.perform",
		"title": "Extract Method"
	},
	"kind": "refactor.extract",
	"title": "Extract Method"
}
```

I think if VS Code is going to make assumptions about things like this, they should be clearly documented so it's clear what other languages should do for the same behaviour. I'm happy to make changes to Dart's server if required, but I'd prefer the behaviour to be documented before doing that (it may be nontrivial).

**Comment by justschen** [comment]
@DanTup I'll give this a look to see if we can support or already have the functionality to do this! just to confirm, even when the label isn't shown, `CtrlCmd + Enter` will not show the preview in the preview panel in the above case for Dart's extract method?

**Comment by justschen** [comment]
the label issue was fixed, but I believe there is discussion ongoing regarding the diff editor and also below ideas to better understand a potential solution !

I think the general idea is that refactorings really should only be classified as `refactorings` if there are edits that are provided. I understand that in some scenarios, a command is provided, and in this case the command leads to edits, but if a code action is providing a command, it doesn't really follow the pattern of our code action types that we have. not 100% convinced this is the right assumption, as you mentioned, and will likely investigate this in the next few weeks.

**Comment by DanTup** [comment]
> I think the general idea is that refactorings really should only be classified as `refactorings` if there are edits that are provided.

I don't understand why this limitation should exist - it prevents any refactor that needs to prompt the user for anything before applying the changes (for example "Move class to File" might prompt for a target file). I understand the current design of commands triggering edits is not convenient, but a fix for that would be to allow the commands to return a `WorkspaceEdit` (something that has been asked for before, because currently with LSP a client can't tell if an edit from a server is tied to a CodeAction/command).

I suspect if this restriction is added, users will just be confused why some portion of their refactors behave differently (because in many cases it won't be possible to change them to edits).

I have opened a PR at https://github.com/microsoft/vscode/pull/210175 that would at least make it possible to have previews on commands (although it wouldn't be consistent with how these refactors work), however like [my other PR that's been open for over a year](https://github.com/microsoft/vscode/pull/175525) it's had zero comments on. I don't know if I have missed some step required to have them looked at 🤷‍♂️

**Comment by justschen** [comment]
Right that makes sense actually! I think if we can tell that there are edits coming, then whether it's a command or not is inconsequential at the end of the day

I'll take a look and give the PR a bump! Appreciate the insight on this!

**Comment by DanTup** [comment]
> Right that makes sense actually! I think if we can tell that there are edits coming, then whether it's a command or not is inconsequential

Yeah, I think the concept of a "command that returns edits" is currently missing and could be useful in a number of places. There's already at least one LSP client ([see here](https://github.com/microsoft/language-server-protocol/issues/1645#issuecomment-1397098508)) that makes assumptions that if the client gets a `workspace/applyEdit` call while it's in the process of executing a command that those things are related.

Maybe on `CodeAction.command` could have a flag to indicate that a command returns edits (instead of triggering an edit out-of-band) and the caller (eg. VS Code's code action handling) executes the command, then applies the returned edits itself.
